src/sources.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    collections::VecDeque,
+    future::Future,
+    sync::{Arc, Mutex},
+};
+use tokio::time::{interval, sleep, Duration};
+
+#[derive(Clone)]
+pub enum Source {
+    Bitfinex,
+    Test(Url, String),
+    CoinGecko,
+    LiveCoinWatch(String), // param is api key
+    CoinMarketCap(String), // param is api key
+}
+
+/**
+ * Wrapper for a request function, for continous attempts, with exponential
+ * backoff.
+ * on_fail is invoked when the request_fn fails, but only if there is any
+ * retries left.
+ */
+async fn request_with_backoff<'a, Fut: 'a, T>(
+    request_fn: impl Fn() -> Fut,
+    on_fail: impl Fn(u64),
+    initial_delay: u64,
+    max_retries: u64,
+) -> Option<T>
+where
+    Fut: Future<Output = Option<T>>, {
+    let mut timeout = initial_delay;
+    let mut retries = max_retries;
+    loop {
+        if let Some(i) = request_fn().await {
+            return Some(i);
+        }
+
+        if retries == 0 {
+            return None;
+        }
+
+        on_fail(timeout);
+
+        retries -= 1;
+        sleep(Duration::from_secs(timeout)).await;
+        timeout *= 2;
+    }
+}
+
+/**
+ * Auxillery function for requesting exchange rate.
+ * Handles common behaviour among functions for requesting exchange rate.
+ * The parser should handle converting the JSON response body into an
+ * exchange rate, and its parameter specifies the expected JSON format.
+ */
+async fn request_exchange_rate_aux<
+    ResponseFormat: for<'de> crypto_common::SerdeDeserialize<'de>,
+>(
+    request: reqwest::RequestBuilder,
+    parser: impl Fn(ResponseFormat) -> Option<f64>,
+    source_label: &str,
+) -> Option<f64> {
+    let resp = match request.send().await {
+        Ok(o) => o,
+        Err(e) => {
+            log::warn!("{}: Unable to send request: {}", source_label, e);
+            return None;
+        }
+    };
+    if resp.status().is_success() {
+        match resp.json::<ResponseFormat>().await {
+            Ok(v) => match parser(v) {
+                Some(val) => {
+                    if val < 0.0 {
+                        log::error!("{}: Exchange rate is negative: {}", source_label, val);
+                        return None;
+                    }
+                    log::debug!("{}: Raw exchange rate CCD in EUR polled: {}", source_label, val);
+                    return Some(val);
+                }
+                None => return None,
+            },
+            Err(err) => {
+                log::error!("{}: Unable to parse response as JSON: {}", source_label, err)
+            }
+        };
+    } else {
+        log::error!("{}: unsuccessful response: {}", source_label, resp.status());
+    };
+    None
+}
+
+/**
+ * Request current exchange rate from bitfinex.
+ */
+async fn request_exchange_rate_bitfinex(client: reqwest::Client) -> Option<f64> {
+    let params = json!({"ccy1": "CCD", "ccy2": "EUR"});
+    let request = client.post(BITFINEX_URL).json(&params);
+    let parser = |v: Vec<f64>| Some(v[0]);
```

Comment:
This will panic for a malformed response. Please handle the error, e.g.,
```suggestion
    let parser = |v: Vec<f64>| v.first().copied();
```